File: app/src/main/java/com/dpis/module/fonts/hookdomain/FontHookDomainPresentation.java
package com.dpis.module.fonts.hookdomain;

import com.dpis.module.R;

import com.dpis.module.hooks.HookDomainOverride;
import com.dpis.module.hooks.HookDomainOverrideStore;

import android.content.Context;

import java.util.Set;

public final class FontHookDomainPresentation {
    private final HookDomainOverride override;

    private FontHookDomainPresentation(HookDomainOverride override) {
        this.override = override != null ? override : HookDomainOverride.automatic();
    }

    public static FontHookDomainPresentation forRecommendedTemplateRaw(String raw) {
        return forOverride(
                HookDomainOverrideStore.fromRaw(raw),
                FontHookDomainRegistry.recommendedTemplateKnownDomains());
    }

    public static FontHookDomainPresentation forOverride(HookDomainOverride override,
            Set<String> automaticKnownDomains) {
        return new FontHookDomainPresentation(
                HookDomainOverrideStore.automaticIfSelectionMatchesAutomatic(
                        override,
                        automaticKnownDomains));
    }

    HookDomainOverride override() {
        return override;
    }

    public boolean displaysAsAutomatic() {
        return !override.customPathEnabled;
    }

    public String normalizedRawOrNull() {
        if (displaysAsAutomatic()) {
            return null;
        }
        return HookDomainOverrideStore.formatCsv(
                override.enabledKnownDomains,
                override.unknownDomains);
    }

    public String buttonText(Context context) {
        if (displaysAsAutomatic()) {
            return context.getString(R.string.dialog_font_hook_domains_title);
        }
        return context.getString(
                R.string.dialog_font_hook_domains_title_with_count,
                selectedDisplayCount(),
                totalDisplayCount());
    }

    public int selectedDisplayCount() {
        return FontHookDomainRegistry.orderedCustomizableDisplaySubset(
                override.enabledKnownDomains).size();
    }

    public int totalDisplayCount() {
        return FontHookDomainRegistry.orderedCustomizableDisplayIdsList().size();
    }
}

File: app/src/main/java/com/dpis/module/fonts/hookdomain/FontHookDomainPresentation.kt
package com.dpis.module.fonts.hookdomain

import android.content.Context
import com.dpis.module.R
import com.dpis.module.hooks.HookDomainOverride
import com.dpis.module.hooks.HookDomainOverrideStore

/** Formats a persisted hook-domain override relative to the automatic baseline. */
class FontHookDomainPresentation private constructor(
    private val domainOverride: HookDomainOverride,
) {
    companion object {
        @JvmStatic fun forAutomaticDomainsRaw(raw: String?) = forOverride(
            HookDomainOverrideStore.fromRaw(raw),
            FontHookDomainRegistry.automaticCustomizableDomains(),
        )

        @JvmStatic fun forOverride(
            domainOverride: HookDomainOverride?,
            automaticKnownDomains: Set<String>?,
        ) = FontHookDomainPresentation(
            HookDomainOverrideStore.automaticIfSelectionMatchesAutomatic(
                domainOverride,
                automaticKnownDomains,
            ),
        )
    }

    fun displaysAsAutomatic() = !domainOverride.customPathEnabled

    fun normalizedRawOrNull(): String? = if (displaysAsAutomatic()) {
        null
    } else {
        HookDomainOverrideStore.formatCsv(domainOverride.enabledKnownDomains, domainOverride.unknownDomains)
    }

    fun buttonText(context: Context): String = if (displaysAsAutomatic()) {
        context.getString(R.string.dialog_font_hook_domains_title)
    } else {
        context.getString(
            R.string.dialog_font_hook_domains_title_with_count,
            selectedDisplayCount(),
            totalDisplayCount(),
        )
    }

    fun selectedDisplayCount() =
        FontHookDomainRegistry.orderedCustomizableDisplaySubset(domainOverride.enabledKnownDomains).size

    fun totalDisplayCount() = FontHookDomainRegistry.orderedCustomizableDisplayIdsList().size
}
